File: app/src/main/java/com/espressif/ui/activities/ProvisionLanding.java
// Copyright 2018 Espressif Systems (Shanghai) PTE LTD
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.espressif.ui.activities;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.os.Vibrator;
import android.provider.Settings;
import android.util.Log;
import android.view.HapticFeedbackConstants;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.widget.ContentLoadingProgressBar;

import com.espressif.AppConstants;
import com.espressif.EspApplication;
import com.espressif.provision.Provision;
import com.espressif.provision.R;
import com.espressif.provision.transport.ResponseListener;
import com.espressif.provision.transport.SoftAPTransport;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;

public class ProvisionLanding extends AppCompatActivity {

    private static final String TAG = "Espressif::" + ProvisionLanding.class.getSimpleName();

    private static final int WIFI_SETTINGS_ACTIVITY_REQUEST = 121;

    private String currentSSID;
    private String deviceNamePrefix;
    private String securityVersion;
    public static SoftAPTransport softAPTransport;
    public static ArrayList<String> deviceCapabilities;
    private SharedPreferences sharedPreferences;

    private TextView tvTitle, tvBack, tvCancel;
    private CardView btnConnect;
    private TextView txtConnectBtn;
    private ImageView arrowImage;
    private ContentLoadingProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_provision_landing);

        sharedPreferences = getSharedPreferences(AppConstants.ESP_PREFERENCES, Context.MODE_PRIVATE);
        deviceNamePrefix = sharedPreferences.getString(AppConstants.KEY_WIFI_NETWORK_NAME_PREFIX, "");
        securityVersion = getIntent().getStringExtra(Provision.CONFIG_SECURITY_KEY);
        deviceCapabilities = new ArrayList<>();

        initViews();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {

        switch (requestCode) {
            case WIFI_SETTINGS_ACTIVITY_REQUEST:
                updateUi();
                break;

            case Provision.REQUEST_PROVISIONING_CODE:
                if (resultCode == RESULT_OK) {
                    setResult(resultCode);
                    finish();
                }
                break;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {

        switch (requestCode) {

            case Provision.REQUEST_PERMISSIONS_CODE: {

                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                }
            }
            break;
        }
    }

    View.OnClickListener btnConnectClickListener = new View.OnClickListener() {

        @Override
        public void onClick(View v) {

            Vibrator vib = (Vibrator) getSystemService(VIBRATOR_SERVICE);
            vib.vibrate(HapticFeedbackConstants.VIRTUAL_KEY);

            deviceNamePrefix = sharedPreferences.getString(AppConstants.KEY_WIFI_NETWORK_NAME_PREFIX, "");
            currentSSID = fetchWifiSSID();

            if (currentSSID != null && currentSSID.startsWith(deviceNamePrefix)) {

                connectDevice();
            } else {
                startActivityForResult(new Intent(Settings.ACTION_WIFI_SETTINGS), WIFI_SETTINGS_ACTIVITY_REQUEST);
            }
        }
    };

    private void connectDevice() {

        String baseUrl = getIntent().getStringExtra(Provision.CONFIG_BASE_URL_KEY);
        softAPTransport = new SoftAPTransport(baseUrl);
        String tempData = "ESP";

        softAPTransport.sendConfigData(AppConstants.HANDLER_PROTO_VER, tempData.getBytes(), new ResponseListener() {

            @Override
            public void onSuccess(byte[] returnData) {

                String data = new String(returnData, StandardCharsets.UTF_8);
                Log.d(TAG, "Value : " + data);

                try {
                    JSONObject jsonObject = new JSONObject(data);
                    JSONObject provInfo = jsonObject.getJSONObject("prov");

                    String versionInfo = provInfo.getString("ver");
                    Log.d(TAG, "Device Version : " + versionInfo);

                    JSONArray capabilities = provInfo.getJSONArray("cap");
                    deviceCapabilities = new ArrayList<>();

                    for (int i = 0; i < capabilities.length(); i++) {
                        String cap = capabilities.getString(i);
                        deviceCapabilities.add(cap);
                    }
                    Log.d(TAG, "Capabilities : " + deviceCapabilities);

                } catch (JSONException e) {
                    e.printStackTrace();
                    Log.d(TAG, "Capabilities JSON not available.");
                }

                if (!deviceCapabilities.contains("no_pop") && securityVersion.equals(Provision.CONFIG_SECURITY_SECURITY1)) {

                    goToPopActivity();

                } else if (deviceCapabilities.contains("wifi_scan")) {

                    goToWifiScanListActivity();

                } else {

                    goToProvisionActivity();
                }
            }

            @Override
            public void onFailure(Exception e) {
                e.printStackTrace();
            }
        });
    }

    private View.OnClickListener cancelButtonClickListener = new View.OnClickListener() {

        @Override
        public void onClick(View v) {

            finish();
        }
    };

    private void initViews() {

        tvTitle = findViewById(R.id.main_toolbar_title);
        tvBack = findViewById(R.id.btn_back);
        tvCancel = findViewById(R.id.btn_cancel);

        tvTitle.setText(R.string.title_activity_connect_device);
        tvBack.setVisibility(View.GONE);
        tvCancel.setVisibility(View.VISIBLE);

        tvCancel.setOnClickListener(cancelButtonClickListener);

        btnConnect = findViewById(R.id.btn_connect);
        txtConnectBtn = findViewById(R.id.text_btn);
        arrowImage = findViewById(R.id.iv_arrow);
        progressBar = findViewById(R.id.progress_indicator);

        txtConnectBtn.setText(R.string.btn_connect);
        btnConnect.setOnClickListener(btnConnectClickListener);
    }

    private void updateUi() {

        ((EspApplication) getApplicationContext()).enableOnlyWifiNetwork();

        currentSSID = fetchWifiSSID();
        Log.d(TAG, "SSID : " + currentSSID);

        if (currentSSID != null && (currentSSID.startsWith(deviceNamePrefix) || currentSSID.equals(deviceNamePrefix))) {

            btnConnect.setEnabled(false);
            btnConnect.setAlpha(0.5f);
            txtConnectBtn.setText(R.string.btn_connecting);
            progressBar.setVisibility(View.VISIBLE);
            arrowImage.setVisibility(View.GONE);
            connectDevice();

        } else {

            btnConnect.setEnabled(true);
            btnConnect.setAlpha(1f);
            txtConnectBtn.setText(R.string.btn_connect);
            progressBar.setVisibility(View.GONE);
            arrowImage.setVisibility(View.VISIBLE);
        }
    }

    private String fetchWifiSSID() {

        ((EspApplication) getApplicationContext()).enableOnlyWifiNetwork();
        String ssid = null;
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {

            ConnectivityManager cm = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo networkInfo = cm.getActiveNetworkInfo();
            if (networkInfo == null) {
                return null;
            }

            if (networkInfo.isConnected()) {
                final WifiManager wifiManager = (WifiManager) getApplicationContext().getSystemService(Context.WIFI_SERVICE);
                final WifiInfo connectionInfo = wifiManager.getConnectionInfo();
                if (connectionInfo != null) {
                    ssid = connectionInfo.getSSID();
                    ssid = ssid.replaceAll("^\"|\"$", "");
                }
            }
        } else {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_COARSE_LOCATION}, Provision.REQUEST_PERMISSIONS_CODE);
        }
        return ssid;
    }

    private void goToPopActivity() {

        finish();
        Intent popIntent = new Intent(getApplicationContext(), ProofOfPossessionActivity.class);
        popIntent.putExtra(AppConstants.KEY_DEVICE_NAME, currentSSID);
        popIntent.putExtras(getIntent());
        startActivity(popIntent);
    }

    private void goToWifiScanListActivity() {

        finish();
        Intent wifiListIntent = new Intent(getApplicationContext(), WiFiScanActivity.class);
        wifiListIntent.putExtras(getIntent());
        wifiListIntent.putExtra(AppConstants.KEY_PROOF_OF_POSSESSION, "");
        startActivity(wifiListIntent);
    }

    private void goToProvisionActivity() {

        finish();
        Intent launchProvisionInstructions = new Intent(getApplicationContext(), ProvisionActivity.class);
        launchProvisionInstructions.putExtras(getIntent());
        launchProvisionInstructions.putExtra(AppConstants.KEY_PROOF_OF_POSSESSION, "");
        startActivityForResult(launchProvisionInstructions, Provision.REQUEST_PROVISIONING_CODE);
    }
}

File: app/src/main/java/com/espressif/ui/activities/ProvisionLanding.kt
// Copyright 2025 Espressif Systems (Shanghai) PTE LTD
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.espressif.ui.activities

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.provider.Settings
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.app.ActivityCompat
import com.espressif.AppConstants
import com.espressif.provisioning.DeviceConnectionEvent
import com.espressif.provisioning.ESPConstants
import com.espressif.ui.utils.Utils
import com.espressif.wifi_provisioning.R
import com.espressif.wifi_provisioning.databinding.ActivityProvisionLandingBinding
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class ProvisionLanding : ManualProvBaseActivity() {

    companion object {
        private val TAG: String = ProvisionLanding::class.java.simpleName

        private const val REQUEST_FINE_LOCATION = 10
        private const val WIFI_SETTINGS_ACTIVITY_REQUEST = 11
    }

    private lateinit var binding: ActivityProvisionLandingBinding

    private var deviceName: String? = null
    private var pop: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityProvisionLandingBinding.inflate(layoutInflater)
        setContentView(binding.root)
        deviceName = intent.getStringExtra(AppConstants.KEY_DEVICE_NAME)
        pop = intent.getStringExtra(AppConstants.KEY_PROOF_OF_POSSESSION)
        initViews()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        when (requestCode) {
            WIFI_SETTINGS_ACTIVITY_REQUEST -> if (hasPermissions()) {
                connectDevice()
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            REQUEST_FINE_LOCATION -> {}
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: DeviceConnectionEvent) {
        Log.d(TAG, "On Device Prov Event RECEIVED : " + event.eventType)

        when (event.eventType) {
            ESPConstants.EVENT_DEVICE_CONNECTED -> {
                Log.d(TAG, "Device Connected Event Received")
                binding.btnConnect.layoutBtn.isEnabled = true
                binding.btnConnect.layoutBtn.alpha = 1f
                binding.btnConnect.textBtn.setText(R.string.btn_connect)
                binding.btnConnect.progressIndicator.visibility = View.GONE
                binding.btnConnect.ivArrow.visibility = View.VISIBLE
                setSecurityTypeFromVersionInfo()
                val isSecure = sharedPreferences!!.getBoolean(AppConstants.KEY_SECURITY_TYPE, true)
                if (isSecure) {
                    if (securityType == AppConstants.SEC_TYPE_0) {
                        Utils.displayDeviceConnectionError(
                            this,
                            getString(R.string.error_security_mismatch)
                        )
                    } else {
                        processDeviceCapabilities()
                    }
                } else {
                    if (securityType != AppConstants.SEC_TYPE_0) {
                        Utils.displayDeviceConnectionError(
                            this,
                            getString(R.string.error_security_mismatch)
                        )
                    } else {
                        processDeviceCapabilities()
                    }
                }
            }

            ESPConstants.EVENT_DEVICE_CONNECTION_FAILED -> {
                binding.btnConnect.layoutBtn.isEnabled = true
                binding.btnConnect.layoutBtn.alpha = 1f
                binding.btnConnect.textBtn.setText(R.string.btn_connect)
                binding.btnConnect.progressIndicator.visibility = View.GONE
                binding.btnConnect.ivArrow.visibility = View.VISIBLE
                Utils.displayDeviceConnectionError(
                    this,
                    getString(R.string.error_device_connect_failed)
                )
            }
        }
    }

    var btnConnectClickListener: View.OnClickListener = View.OnClickListener {
        startActivityForResult(
            Intent(Settings.ACTION_WIFI_SETTINGS),
            WIFI_SETTINGS_ACTIVITY_REQUEST
        )
    }

    private fun connectDevice() {
        binding.btnConnect.layoutBtn.isEnabled = false
        binding.btnConnect.layoutBtn.alpha = 0.5f
        binding.btnConnect.textBtn.setText(R.string.btn_connecting)
        binding.btnConnect.progressIndicator.visibility = View.VISIBLE
        binding.btnConnect.ivArrow.visibility = View.GONE

        if (ActivityCompat.checkSelfPermission(
                this@ProvisionLanding,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            provisionManager!!.espDevice.connectWiFiDevice()
        } else {
            Log.e(TAG, "Not able to connect device as Location permission is not granted.")
            Toast.makeText(
                this@ProvisionLanding,
                "Please give location permission to connect device",
                Toast.LENGTH_LONG
            ).show()
        }
    }

    private fun initViews() {
        setToolbar()
        var instruction = getString(R.string.connect_device_instruction_general)
        
        if (TextUtils.isEmpty(deviceName)) {
            binding.tvConnectDeviceInstruction.text = instruction
            binding.tvDeviceName.visibility = View.GONE
        } else {
            instruction = getString(R.string.connect_device_instruction_specific)
            binding.tvConnectDeviceInstruction.text = instruction
            binding.tvDeviceName.visibility = View.VISIBLE
            binding.tvDeviceName.text = deviceName
        }

        binding.btnConnect.textBtn.setText(R.string.btn_connect)
        binding.btnConnect.layoutBtn.setOnClickListener(btnConnectClickListener)
        hasPermissions()
    }

    private fun setToolbar() {
        setSupportActionBar(binding.titleBar.toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        supportActionBar!!.setTitle(R.string.title_activity_connect_device)
        binding.titleBar.toolbar.navigationIcon =
            AppCompatResources.getDrawable(this, R.drawable.ic_arrow_left)
        binding.titleBar.toolbar.setNavigationOnClickListener {
            finish()
        }
    }

    private fun processDeviceCapabilities() {
        val deviceCaps = provisionManager!!.espDevice.deviceCapabilities

        if (deviceCaps != null) {
            if (!TextUtils.isEmpty(pop)) {
                provisionManager!!.espDevice.proofOfPossession = pop

                if (deviceCaps.contains(AppConstants.CAPABILITY_WIFI_SCAN)) {
                    goToWifiScanListActivity()
                } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_SCAN)) {
                    goToThreadScanActivity(true)
                } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_PROV)) {
                    goToThreadScanActivity(false)
                } else {
                    goToWiFiConfigActivity()
                }
            } else {
                if (!deviceCaps.contains("no_pop") && securityType != AppConstants.SEC_TYPE_0) {
                    goToPopActivity()
                } else if (deviceCaps.contains(AppConstants.CAPABILITY_WIFI_SCAN)) {
                    goToWifiScanListActivity()
                } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_SCAN)) {
                    goToThreadScanActivity(true)
                } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_PROV)) {
                    goToThreadScanActivity(false)
                } else {
                    goToWiFiConfigActivity()
                }
            }
        } else {
            goToWiFiConfigActivity()
        }
    }

    private fun goToPopActivity() {
        finish()
        val popIntent = Intent(applicationContext, ProofOfPossessionActivity::class.java)
        startActivity(popIntent)
    }

    private fun goToWifiScanListActivity() {
        finish()
        val wifiListIntent = Intent(applicationContext, WiFiScanActivity::class.java)
        startActivity(wifiListIntent)
    }

    private fun goToThreadScanActivity(scanCapAvailable: Boolean) {
        finish()
        val threadConfigIntent = Intent(applicationContext, ThreadConfigActivity::class.java)
        threadConfigIntent.putExtra(AppConstants.KEY_THREAD_SCAN_AVAILABLE, scanCapAvailable)
        startActivity(threadConfigIntent)
    }

    private fun goToWiFiConfigActivity() {
        finish()
        val wifiConfigIntent = Intent(applicationContext, WiFiConfigActivity::class.java)
        startActivity(wifiConfigIntent)
    }

    private fun hasPermissions(): Boolean {
        if (!hasLocationPermissions()) {
            requestLocationPermission()
            return false
        }
        return true
    }

    private fun hasLocationPermissions(): Boolean {
        return checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestLocationPermission() {
        requestPermissions(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), REQUEST_FINE_LOCATION)
    }
}
